charter.go

```diff
@@ -0,0 +1,90 @@
+package parser
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/Masterminds/sprig/v3"
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	templateDir       = "template"
+	templateExtension = ".gotmpl"
+)
+
+type ChartData struct {
+	Config  *Config
+	Version string
+	Ports   []*ContainerPort
+}
+
+func (data *ChartData) Chart() (map[string][]byte, error) {
+	return data.recursiveChart(templateDir)
+}
+
+func (data *ChartData) recursiveChart(dirName string) (map[string][]byte, error) {
+	outFiles := make(map[string][]byte)
+
+	dir, err := os.ReadDir(dirName)
+	if err != nil {
+		return nil, fmt.Errorf("failed reading directory %s: %w", dirName, err)
+	}
+
+	for _, entry := range dir {
+		newPath := filepath.Join(dirName, entry.Name())
+		if entry.IsDir() {
+			newFiles, err := data.recursiveChart(newPath)
+			if err != nil {
+				return nil, err
+			}
+
+			for p, data := range newFiles {
+				outPath := filepath.Join(newPath, p)
+				if strings.HasPrefix(outPath, templateDir) {
+					outPath = strings.TrimLeft(outPath, templateDir)
+				}
+				outFiles[outPath] = data
+			}
+
+			continue
+		}
+
+		file, err := os.ReadFile(newPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed reading file %s: %w", newPath, err)
+		}
+
+		if filepath.Ext(entry.Name()) == templateExtension {
+			funcMap := sprig.HermeticTxtFuncMap()
```

Comment:
This logic should probably be outside of the directory walking code and in another function